### ISSUE TITLE
Fundamental design of the cube processes (reduce, filter, apply, aggregate, ...)

### ISSUE THREAD

**Issue by jdries** [Open-EO/openeo-processes#16]
In the process graph example given in this issue:
https://github.com/Open-EO/openeo-api/issues/160
2 uses of the reduce operation are shown:
1. a reduce over the time dimension using the min operator
2. a reduce over the spectral dimension, to compute an NDVI
I fully get the first example, but the second one is a more special case.
It uses substraction and division operators, which assume that the nir and red bands are stored in the right order, this is however not clearly enforced by the process graph.
Also, if we look at a general definition of a reduction operator:
https://en.wikipedia.org/wiki/Reduction_Operator
substraction and division are not really usable as generic reduction operators.

Basically it seems that what we need is a process that applies a complex function to all values of a particular dimension, and returns a new set of values for the same dimension:
spectral_bands' = f(spectral_bands)
Almost like apply_dimension, but in that case the output shape has to be the same as the input shape.

**Comment by mkadunc** [comment]
In 2), the bands are enforced by a previous filter operation:
```
.filter_bands(bands: ["nir", "red"]) 	//filter3
.reduce(dimension: "spectral", reducer: new callback({ 	//reduce2
    temp = args["dimension_data"]
    divide1 = divide(x: temp.subtract(), y: temp.sum())
    output1 = divide1.output()
    return divide1
}))
```

You're right that `reduce` may be improperly named - in JavaScript terms, our processes would be:
- apply_dimensions: [Array.map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) with adjustable direction/depth, requiring same shape of output
- reduce: [Array.map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) with adjustable direction/depth, without requirement for same shape of output
- aggregate: [Array.reduce](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce) - this is what *reduce* generally means - iteratively accumulate some data into the aggregated result using a *reducer* (in Javascript, reducer adds a single element of the iteration to the previously accumulated data; with another *merger*, you can do a tree-like aggregation (see https://en.wikipedia.org/wiki/Fold_(higher-order_function) )

**Comment by jdries** [comment]
filter_bands also does not enforce an ordering, or that's not (yet) documented.
Note that for aggregate/reduce, it would be ideal if the backends do not have to provide guarantees on the order in which the function is applied, so we can process our inputs in parallel. 

In numpy, the [apply_along_axis](https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.apply_along_axis.html) function would be used for the reduce/NDVI case.

**Comment by m-mohr** [comment]
`filter_bands` was planned to enforce the order, but at some point got lost. I'll add a remark to the description. Nevertheless, I'm not sure whether that is really the best solution to rely on the actual order of the bands. Is there a better alternative?

> Almost like apply_dimension, but in that case the output shape has to be the same as the input shape.

Maybe we use different words here, so what do you mean by shape?

Regarding naming: I'm open to change names if there are better proposals, please let us know.

**Comment by mkadunc** [comment]
> Maybe we use different words here, so what do you mean by shape?

shape of the data cube is everything but its values - i.e. ordered axes with their domains, limits (min/max) and other metadata, variable names

**Comment by m-mohr** [comment]
We'll revisit this in the dev telco next week, once we have the new processes defined for #18 as a proposal in the API.

**Comment by m-mohr** [comment]
Okay, I'll give it a shot here to come up with names (and more) for our fundamental [higher-order](https://en.wikipedia.org/wiki/Higher-order_function) cube processes inspired from functional programming. Currently they are named filter*, aggregate* (+ resample*), reduce, apply* + some we planned to add in the last telco.

**Definitions:**

* Unary process = one input value, one return value
* Binary process = two inputs values, one return value
* n-ary process = multiple / a list of input values, multiple / a list of return values
* reducer process = two inputs or a list of input values, but returns potentially less values (usually a single value)

**Requirements:**

First of all, we need find out what kind of processes we (probably) need:

- Execute a unary process on each value (currently `apply`)
- Execute a n-ary process on all values (currently `apply_dimension`)
- Get selected values based on a logical expression (= unary process) (currently `filter`)
- Reduce a list of values to lower cardinality (usually *n* to *1*) - Can be defined as:
  - Reducer working with binary processes (MapReduce, currently N/A)
  - Reducer working with n-ary processes (currently `reduce`)
- Compute new values based on "groups" (currently `aggregate`) and/or increase/decrease resolution (currently `resample`).

**Names:**

Regarding the naming of functions there are nice lists on Wikipedia, which compare the names for many programming languages:

* [Filter (higher-order function)](https://en.wikipedia.org/wiki/Filter_(higher-order_function)):
  Commonly used is `filter`, some use `select`, rarely used are where, find and others.
* [Map (higher-order function)](https://en.wikipedia.org/wiki/Map_(higher-order_function)):
  Commonly used is `map`, some use `apply`, rarely used are collect, filter and others
* [Fold (higher-order function)](https://en.wikipedia.org/wiki/Fold_(higher-order_function)):
  Commonly used are `reduce` and `fold` (some use suffixes such as `r` and `l` for right and left fold), rarely used are inject, aggregate and others. I'm currently only considering left fold, for right fold we 
   might consider adding a `reverse` function or so later.

**Discussion based on the requirements:**

1. *Execute a unary process on each value*

    Currently, we have `apply` for this, but it is lacking a `dimension` parameter, isn't it? I don't see much use in applying a process on all values in all dimensions, usually it would be applying a unary process on one or more selected dimensions. So, I propose to add an optional parameter `dimension` to apply (similar to the one in `apply_dimension`). (We could also rename `apply` to `map`, but I personally think apply is easier to grasp than map.)

    I expect that the apply process always returns as many elements as it received. Recent discussion in #18 by @mkadunc though doesn't necessarily want this constraint for this process as far as I understood? (Same applies for apply_dimension).

2. *Execute a n-ary process on all values*

    This process is basically a workaround to allow that our processes such as `sum` are generalized for arrays and not depend on data cubes and their dimensions. GEE for example just duplicates the processes for every data type (e.g. array, image collection, feature collection, ...). Seems okay as it is right now, but I am not so happy with the name any longer. The only difference to `apply` is that  we work on n-ary processes instead of unary processes so the `_dimension` suffix may be inappropriate. What could be an alt6ernative? A first quick idea was to have `apply_each` (instead of `apply`) and `apply_all`. 

3. *Get selected values based on a logical expression*

    `filter` seems to be okay as it is.

4. *Reduce a list of values, can be defined for binary processes (MapReduce) and/or n-ary processes*

    Currently, everything works on lists. This may not scale well (but could be optimized by the back-end to internally use map-reduce) and doesn't allow passing a UDF which expects two inputs. Also, some people are confused by reducers that are not working on binary input. Nevertheless, it allows to compute exact `median` and `quantiles`, which may not always be required.

    If we would like to switch to binary processes as callbacks this would probably require to rewrite many of the processes that are working on arrays at the moment, e.g. `sum`, `count`, `mean`, `median`, `last` etc. They would be redefined to work on two inputs instead of a list, e.g.  `min(a, b)` instead of `min(list)`. For some of the processes I'm not quite sure how we'll define them in a MapReduce environment (e.g. `median`, `quantiles`). 

    Make the `reduce` callback work on binary processes. In this case we would need binary processes, e.g. `min(a, b)` instead of `min(list)`. How to specify `median` and `quantile`?

    Could we have two processes, one working binary one on lists? Would that confuse users? Regarding names we could go with pairs such as `reduce` / `map_reduce` or names could be inspired by how we name apply / apply_dimension.

5. Compute new values based on "groups" and/or increase/decrease resolution

    We have `aggregate_*` and `resample_*` at the moment. They probably need some fine-tuning (see #26, #14). The names are quite intuitive for me personally, but there seems to be much confusion around these terms, which will probably never be solved (by us). So we need to make sure they are well documented.

Thoughts? Opinions? We haven't quite solved it yet. Especially the reducers give me a headache.

**Comment by jdries** [comment]
A reduction operator is often defined as a commutative and associative operator:
https://en.wikipedia.org/wiki/Reduction_Operator
If you use this definition, the problem of using list input versus binary input goes away: the backend can provide the arguments as a list of 2 or more elements, but the result will always be the same.

For median, I would not mind having 'special' reduce operators like GEE:
`var median = collection.reduce(ee.Reducer.median());`
Where the backend will have to provide custom, non-generic implementations.

**Comment by mkadunc** [comment]
I think the distinction between *unary*, *binary* and *n-ary* is not necessary in this context, as it distracts from the point of what the higher-order functions actually do and what the differences between them are. Furthermore, if I used *arity* in this discussion, I would not define a function that takes a zero-dimensional cube *unary*, and one that takes a 1-dimensional cube *n-ary*. If we need arity, i'd define it as this:
- *unary* function/process takes one cube (of any shape) as its input and returns one cube as its output (output can be any shape, unrelated to shape of input)
- *binary* function/process takes two cubes (of any shape) as its input and returns one cube (of any shape)
- *n-ary* function/process takes many cubes (of any shape) as its input and returns one cube (of any shape) as output

The higher-order functions differ not so much in the arity of their callback function, but in what they do to the input and how they achieve this.

**Comment by m-mohr** [comment]
We discussed several related things in the telco today:
- There were different versions of data cubes in our minds and that was very confusing. We need a glossary for data cube terms, @edzer will write one. (data cube, dimension, dimension value, pixel value, ...)
- We discussed bands with the conclusion that they are a dimension, e.g. "spectral", but potentially with different names etc for other types of data such as SAR data.
- We discussed about a band arithmetic process again and haven't decided yet to add one. Could be added later if really necessary.
- apply and apply_dimension are good as they are, but probably need a better description what you can pass as callbacks (unary vs n-ary). Descriptions should explicitly say that the values you work with are the pixel values.
- Open point with a tendency to "not needed (yet)": Do we need a way to change/iterate over dimension labels?
- resample: See #14 
- reduce: See #18 
- filters: Needs to explicitly say that the values you compare against are the dimension values.
- Haven't much spoken about names, nobody really complained about them so much in the telco so I'll stick with them until further issues/comments pop up.
- add_dimension/reduce: A question was how we can inform a datacube what type of dimension we add, e.g. that a new dimension is temporal. We could add a parameter to set a type (and need to define types). Use case was that a reduce could create a new temporal dimension and therefore we need to specify a new dimension and the type of it. We could then let the user choose a target dimension in the process.
- geometry related processes: see #26

**Comment by mkadunc** [comment]
> There were different versions of data cubes in our minds and that was very confusing. We need a glossary for data cube terms, @edzer will write one. (data cube, dimension, dimension value, pixel value, ...)

When we do write this glossary - the filter/map/reduce/aggregate definitions and discussions would also benefit from an unanbiguous definition of:

- *groups* (which @m-mohr mentions in 5.) and/or *sub-cubes* in *partitions* - there are many ways in which a user may want to split a data cube into a multitude of sub-cubes, and this splitting very heavily informs all of the higher-order functions, aggregate_* functions and maybe even zonal statistics

- *index* and *multiindex* or similar: for unambiguous definitions and discussions surrounding the behaviour of higher-order functions we'd benefit form an agreed terminology on addressing a single cell of a datacube, a single plane, or some other kind of subcube (or maybe even a group in an arbitrary partition).

- *variables*, *properties*, *metadata* - I'm still not sure whether we support many-variables-per-cube or do we achieve the equivalent functionality by having many data cubes, one per each variable. Note that in NetCDF, even coordinate labels and positions are variables, so even if we decide to have single-variable cubes we might want to keep the option of having 1D variables (maybe that's what's meant by dimension value?)

I have a feeling it would also help the discussion if we all adopted a suitable (mathematical) notation that would help us discuss the mechanics of various processes more clearly,

**Comment by m-mohr** [comment]
@edzer There are some more suggestions for terms that could need some clarification in the glossary.

**Comment by edzer** [comment]
Draft glossary is here: https://github.com/edzer/openeo-api/blob/0.4.0/docs/glossary.md
Coming back to @mkadunc comments above:

> When we do write this glossary - the filter/map/reduce/aggregate definitions and discussions would also benefit from an unanbiguous definition of:
> * _groups_ (which @m-mohr mentions in 5.) and/or _sub-cubes_ in _partitions_ - there are many ways in which a user may want to split a data cube into a multitude of sub-cubes, and this splitting very heavily informs all of the higher-order functions, aggregate_* functions and maybe even zonal statistics

I mention these in `aggregate`. We don't have a grouped data cube as explicit object or result. I don't know a use case where we'd need them otherwise than applying a reducer to each group, which is what `aggregate` does.

> * _index_ and _multiindex_ or similar: for unambiguous definitions and discussions surrounding the behaviour of higher-order functions we'd benefit from an agreed terminology on addressing a single cell of a datacube, a single plane, or some other kind of subcube (or maybe even a group in an arbitrary partition).

We don't have this at the moment. We don't provide integer indexing. `filter` returns subcubes, based on expressions containing dimension values.

> * _variables_, _properties_, _metadata_ - I'm still not sure whether we support many-variables-per-cube or do we achieve the equivalent functionality by having many data cubes, one per each variable. Note that in NetCDF, even coordinate labels and positions are variables, so even if we decide to have single-variable cubes we might want to keep the option of having 1D variables (maybe that's what's meant by dimension value?)

We support single variable (grayscale) pixels. If we would have variables in a cube, we'd need a second layer of semantics to address them. Currently we don't have the infrastructure to handle dimensions as variables, but I can see the advantage of having this. But do we need it for the use cases?

Properties and metadata: my suggestion (to be discussed tomorrow) is to not treat these as dimensions, but to filter on them when we create a data cube from an image collection (`load_collection`). This avoids having to carry them around through all the cube logic.

> I have a feeling it would also help the discussion if we all adopted a suitable (mathematical) notation that would help us discuss the mechanics of various processes more clearly,

I agree, in particular in a language that everyone understands.

**Comment by m-mohr** [comment]
> > I have a feeling it would also help the discussion if we all adopted a suitable (mathematical) notation that would help us discuss the mechanics of various processes more clearly,
> 
> I agree, in particular in a language that everyone understands.

So, who is doing that? Probably more targeted towards a 0.5 release than for the 0.4 though...

**Comment by edzer** [comment]
Maybe we can discuss the language first. I'd be happy to do it in R, that is in any case both widely understood, open source, and executable. But I may be the only one with that preference? What kind of mathematical notations were you thinking of, @mkadunc ?

**Comment by m-mohr** [comment]
I think all issues raised here are solved now. Closing.
